src/lib/falcor/routes/articles/by-id.js

```diff
@@ -0,0 +1,115 @@
+import falcor from 'falcor';
+import _ from 'lodash';
+
+import {
+  articleQuery,
+  articleAuthorQuery,
+  updateArticles,
+  updateAuthors,
+} from 'lib/db';
+
+import { cleanupJsonGraphArg } from 'lib/falcor/falcor-utilities';
+import { has } from 'lib/utilities';
+
+const $ref = falcor.Model.ref;
+
+export default [
+  {
+    // Get custom article data from MariaDB
+    route:
+      "articles['byId'][{keys:ids}]['id', 'image_url', 'slug', 'title', 'markdown', 'html', 'teaser', 'category', 'published_at', 'views', 'is_interactive']", // eslint-disable-line max-len
+    get: async pathSet => {
+      const requestedFields = pathSet[3];
+      const data = await articleQuery('id', pathSet.ids, requestedFields);
+      const results = data
+        .map(article => {
+          const processedArticle = { ...article };
+          if (
+            has.call(processedArticle, 'published_at') &&
+            processedArticle.published_at instanceof Date
+          ) {
+            processedArticle.published_at = processedArticle.published_at.getTime();
+          }
+          return requestedFields.map(field => ({
+            path: ['articles', 'byId', processedArticle.id, field],
+            value: processedArticle[field],
+          }));
+        })
+        .flatten();
+      return results;
+    },
+    set: async jsonGraphArg => {
+      jsonGraphArg = cleanupJsonGraphArg(jsonGraphArg); // eslint-disable-line no-param-reassign
+      const articles = jsonGraphArg.articles.byId;
+      const flag = await updateArticles('id', articles);
+      if (!flag) {
+        throw new Error(
+          'For unknown reasons updatePostMeta returned a non-true flag',
+        );
+      }
+      const results = _.map(articles, (singleArticle, id) =>
+        _.map(singleArticle, (value, field) => ({
+          path: ['articles', 'byId', id, field],
+          value,
+        })),
+      ).flatten();
+      return results;
+    },
+  },
+  {
+    // Get author information from article
+    route: "articles['byId'][{keys:ids}]['authors'][{integers:indices}]",
+    get: pathSet =>
+      new Promise(resolve => {
+        articleAuthorQuery('id', pathSet.ids).then(data => {
+          // We receive the data as an object with keys equalling article ids
+          // and values being an array of author ids in no particular order
+          const results = [];
+          _.forEach(data, (authorSlugArray, postId) => {
+            pathSet.indices.forEach(index => {
+              if (index < authorSlugArray.length) {
+                results.push({
+                  path: ['articles', 'byId', postId, 'authors', index],
+                  value: $ref(['staff', 'bySlug', authorSlugArray[index]]),
```

Comment:
Note that here, in order to be able to change slugs on authors, this should be refactored to reference author by id.